### PR TITLE
evaluator, parser: support UTC_DATE()

### PR DESCRIPTION
**Pull request by overvenus**
see issue: #236.

### CODE REVIEW

**Comment by shenli** [comment]
@overvenus Thanks for your report!  We will add more builtin functions. 
If you are interested in TiDB, would you please send us a PR? We have a tutorial about how to add a new builtin function. We will be appreciated.

**Comment by overvenus** [comment]
@shenli Hi, and yes, I have just added a new builtin function, it is a PR not issue. Maybe I should change title to **Add UTC_DATE()**?

**Comment by shenli** [comment]
Sorry.... It is my mistake. 
Great! I will take a look!

File: evaluator/builtin_time_test.go [hunk 1]
```diff
@@ -316,6 +316,14 @@ func (s *testEvaluatorSuite) TestCurrentTime(c *C) {
 	c.Assert(err, NotNil)
 }
 
+func (s *testEvaluatorSuite) TestUTCDate(c *C) {
```
Comment (shenli) [inline evaluator/builtin_time_test.go]: Can you add another case like:
mysql> select utc_date + 1;
+--------------+
| utc_date + 1 |
+--------------+
|     20160402 |
+--------------+

Comment (overvenus) [inline evaluator/builtin_time_test.go]: ok, I will take a look, and maybe modification of current_date() is also needed.

File: evaluator/builtin_time.go [hunk 2]
```diff
@@ -390,6 +390,16 @@ func builtinCurrentTime(args []types.Datum, _ context.Context) (d types.Datum, e
 	return convertToDuration(d, fsp)
 }
 
+// See https://dev.mysql.com/doc/refman/5.7/en/date-and-time-functions.html#function_utc-date
+func builtinUTCDate(args []types.Datum, _ context.Context) (d types.Datum, err error) {
+	year, month, day := time.Now().Date()
+	t := mysql.Time{
+		Time: time.Date(year, month, day, 0, 0, 0, 0, time.Local),
```
Comment (shenli) [inline evaluator/builtin_time.go]: time.Local -> time.UTC ?

Comment (overvenus) [inline evaluator/builtin_time.go]: @shenli It does not work.

time.UTC:
```
mysql> select utc_date+1;
+----------------+
| utc_date+1     |
+----------------+
| 20160401000001 |
+----------------+
1 row in set (0.00 sec)
```

And please take a look **current_date**:

**TiDB**:
```
mysql> select utc_date+1, current_date+1;
+----------------+----------------+
| utc_date+1     | current_date+1 |
+----------------+----------------+
| 20160401000001 | 20160401000001 |
+----------------+----------------+
1 row in set (0.00 sec)
```

**MySQL**:

```
mysql> select utc_date+1, current_date+1;
+------------+----------------+
| utc_date+1 | current_date+1 |
+------------+----------------+
|   20160402 |       20160402 |
+------------+----------------+
1 row in set (0.00 sec)
```

File: evaluator/builtin_time_test.go [hunk 3]
```diff
@@ -316,6 +316,14 @@ func (s *testEvaluatorSuite) TestCurrentTime(c *C) {
...
+	last := time.Now()
+	v, err := builtinUTCDate(types.MakeDatums(nil), nil)
+	c.Assert(err, IsNil)
+	n := v.GetMysqlTime()
+	c.Assert(n.String(), GreaterEqual, last.Format(mysql.DateFormat))
```
Comment (shenli) [inline evaluator/builtin_time_test.go]: last is local time, but utcDate is utc time. So last should be time.Now().UTC() ?

File: evaluator/builtin_time.go (hunk 2 above)
Comment (overvenus) [inline evaluator/builtin_time.go]: https://github.com/pingcap/tidb/blob/master/util/types/datum.go#L621

Probably, we need a new `case mysql.TypeDate:`

Comment (shenli) [inline evaluator/builtin_time.go]: No. When evaluating currend_date()+1,  we should convert date to int. So the target type is int: https://github.com/pingcap/tidb/blob/master/util/types/datum.go#L615

The problem is when convert date to int, we should drop the hour/minute/second.
Maybe we should fix it here: https://github.com/pingcap/tidb/blob/master/mysql/time.go#L223

**Comment by shenli** [comment]
@coocood @qiuyesuifeng PTAL

File: evaluator/builtin_time.go (hunk 2 above)
Comment (overvenus) [inline evaluator/builtin_time.go]: Well, we should drop the hour/minute/second before convert date to int, but fix it in [ToNumber](https://github.com/pingcap/tidb/blob/master/mysql/time.go#L223), I dont think its a good idea.

Comment (shenli) [inline evaluator/builtin_time.go]: Date/Datetime/timestamp use the same data structure mysql/time.go. So when it store a date value, the hour/minute/second are all zero. We can not drop them before.
tidb> select cast(current_date() as SIGNED);
+--------------------------------+
| cast(current_date() as SIGNED) |
+--------------------------------+
| 20160402000000                 |
+--------------------------------+
1 row in set (0.01 sec)

mysql> select cast(current_date() as SIGNED);
+--------------------------------+
| cast(current_date() as SIGNED) |
+--------------------------------+
|                       20160402 |
+--------------------------------+
1 row in set (0.00 sec)

It is OK to raise a new issue and fix this bug in another PR. We can focus on UTC time in this PR.

Comment (qiuyesuifeng) [inline evaluator/builtin_time.go]: Agreed, should be in next PR.

**Comment by shenli** [comment]
@overvenus Any update for this PR?

**Comment by overvenus** [comment]
@shenli @qiuyesuifeng PTAL

**Comment by shenli** [comment]
LGTM
@coocood @qiuyesuifeng

**Comment by qiuyesuifeng** [comment]
@overvenus  Thanks.
LGTM

**Comment by overvenus** [comment]
Squashed. @shenli @qiuyesuifeng 

cc @c4pt0r

File: evaluator/builtin_time.go [hunk 4]
```diff
@@ -390,6 +390,16 @@ func builtinCurrentTime(args []types.Datum, _ context.Context) (d types.Datum, e
 	return convertToDuration(d, fsp)
 }
 
+// See https://dev.mysql.com/doc/refman/5.7/en/date-and-time-functions.html#function_utc-date
+func builtinUTCDate(args []types.Datum, _ context.Context) (d types.Datum, err error) {
+	year, month, day := time.Now().Date()
```
Comment (shenli) [inline evaluator/builtin_time.go]: Maybe we should change this to time.Now().UTC().Date()?

Comment (overvenus) [inline evaluator/builtin_time.go]: Oops, I'm sorry, you are right, thank you for pointing out, I will fix it tomorrow.

**Comment by overvenus** [comment]
PTAL

**Comment by shenli** [comment]
LGTM

**Pull request merged**